e2e/rapid_test.go

```diff
@@ -0,0 +1,598 @@
+	wg := errgroup.Group{}
+	for i := range cluster {
+		i := i
+		state := cluster[i].GetState()
+		isLocked := cluster[i].IsLocked()
+		wg.Go(func() (err1 error) {
+			wgTime := time.Now()
+			exitCh := make(chan struct{})
+			deadlineTimeout := waitDuration
+			deadline := time.After(deadlineTimeout)
+			if stuckList.Get(i) {
+				return nil
+			}
+
+			if doneList.Get(i) {
+				return nil
+			}
+
+			go func() {
+				stuckList.Set(i, true)
+				defer func() {
+					stuckList.Set(i, false)
+				}()
+				cluster[i].RunCycle(context.Background())
+				close(exitCh)
+			}()
+			select {
+			case <-exitCh:
+			case <-deadline:
+			}
+
+			//useful for debug
+			_, _, _ = state, wgTime, isLocked
+			//if time.Since(wgTime) > waitDuration {
+			//	fmt.Println("wgitme ", state, i, callNumber, time.Since(wgTime), err1, isLocked)
+			//}
+			//
+			return err1
+		})
+	}
+
+	return wg.Wait()
+}
+
+func setDoneOnDoneState(cluster []*pbft.Pbft, doneList *BoolSlice) {
+	for i, node := range cluster {
+		state := node.GetState()
+		if state == pbft.DoneState {
+			doneList.Set(i, true)
+		}
+	}
+}
+
+func maxPredefinedRound(mp map[uint64]map[uint64][]uint64) uint64 {
+	var max uint64
+	for i := range mp {
+		if i > max {
+			max = i
+		}
+	}
+	return max
+}
+
+func sendTimeoutIfNNodesStucked(timeoutsChan []chan time.Time, numOfNodes int) func(stuckList *BoolSlice) bool {
+	return func(stuckList *BoolSlice) bool {
+		if stuckList.CalculateNum(true) == numOfNodes {
+			for i := range timeoutsChan {
+				timeoutsChan[i] <- time.Now()
+			}
+		}
+		return false
+	}
+}
+
+func runCluster(ctx context.Context,
+	cluster []*pbft.Pbft,
+	handleStuckList func(*BoolSlice) bool,
+	handleDoneList func(*BoolSlice) bool,
+	handleMaxRoundNumber func(uint64) bool,
+	limitCallNumber int,
+) error {
+	for i := range cluster {
+		cluster[i].SetInitialState(context.Background())
+	}
+
+	stuckList := NewBoolSlice(len(cluster))
+	doneList := NewBoolSlice(len(cluster))
+
+	callNumber := 0
+	for {
+		callNumber++
+		err := runClusterCycle(cluster, callNumber, stuckList, doneList)
+		if err != nil {
+			return err
+		}
+
+		setDoneOnDoneState(cluster, doneList)
+		if handleStuckList(stuckList) {
+			return nil
+		}
+		if handleDoneList(doneList) {
+			return nil
+		}
+
+		if handleMaxRoundNumber(getMaxClusterRound(cluster)) {
+			return nil
+		}
+		if callNumber > limitCallNumber {
+			return errors.New("callnumber limit")
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+
+		}
+	}
+}
+
+//generateRoundsRoutingMap generator for routing map
+//map[uint64]map[uint64][]uint64
+func generateRoundsRoutingMap(numOfNodes uint64) *rapid.Generator {
+	return rapid.MapOfN(
+		//round number
+		rapid.Uint64Range(0, 3),
+		//validator routing
+		generateRoutingMap(numOfNodes, 0),
+		//min num of rounds
+		1,
+		//max num of rounds
+		10)
+}
+
+func generateRoutingMap(numberOfNode uint64, minNumberOfConnections int) *rapid.Generator {
```

Comment:
If we are going to remove `generateRoundsRoutingMap`, we don't need this one either.